### PR TITLE
Nexus command processing

### PR DESCRIPTION
**Pull request by bergundy**
## What changed?

- Added Nexus command processing logic
- Created abstractions to allow a plugin (to be renamed component) to register command handlers
- Added a generic `AddHistoryEvent` helper to mutable state to start breaking up the logic from mutable state monolith into the components.

## Why?

This is part of the work to enable starting an operation from a workflow.

## How did you test it?

Added unit and functional tests.

## Potential risks

Replication isn't addressed yet, this code shouldn't make it into production.
Note that this is stacked commit on top of #5545 that is targeting the `nexus` feature branch. I'll make sure to put dynamic configs in front of all of this unfinished functionality.

### CODE REVIEW

File: service/history/command_checker.go [hunk 1]
```diff
@@ -943,14 +943,17 @@ func (v *commandAttrValidator) validateCommandSequence(
...
+			enumspb.COMMAND_TYPE_SCHEDULE_NEXUS_OPERATION,
+			enumspb.COMMAND_TYPE_REQUEST_CANCEL_NEXUS_OPERATION:
 			// noop
 		case enumspb.COMMAND_TYPE_CONTINUE_AS_NEW_WORKFLOW_EXECUTION,
 			enumspb.COMMAND_TYPE_COMPLETE_WORKFLOW_EXECUTION,
 			enumspb.COMMAND_TYPE_FAIL_WORKFLOW_EXECUTION,
 			enumspb.COMMAND_TYPE_CANCEL_WORKFLOW_EXECUTION:
 			closeCommand = command.GetCommandType()
 		default:
+			// Why is this the default? Can't we assume that any specified command is by default not a close command?
```
Comment (yycptt) [inline service/history/command_checker.go:956]: probably because it can force you to update this logic when adding a close command.

File: service/history/workflow/mutable_state_impl.go [hunk 2]
```diff
@@ -424,6 +425,10 @@ func NewMutableStateFromDB(
 		}
 	}
 
+	// Proto serialization replaces empty maps with nils, ensure this map is never nil.
+	if mutableState.executionInfo.SubStateMachinesByType == nil {
+		mutableState.executionInfo.SubStateMachinesByType = make(map[int32]*persistencespb.StateMachineMap)
+	}
```
Comment (yycptt) [inline service/history/workflow/mutable_state_impl.go:431]: This should be done by the persistence serializer. We have something similar in the ShardInfoFromBlob method.

File: service/history/workflow/mutable_state_impl.go [hunk 3]
```diff
@@ -765,6 +770,14 @@ func (ms *MutableStateImpl) GetNamespaceEntry() *namespace.Namespace {
 	return ms.namespaceEntry
 }
 
+// AddHistoryEvent adds any history event to this workflow execution.
+// The provided setAttributes function should be used to set the attributes on the event.
+func (ms *MutableStateImpl) AddHistoryEvent(t enumspb.EventType, setAttributes func(*history.HistoryEvent)) *history.HistoryEvent {
+	event := ms.hBuilder.AddHistoryEvent(t, setAttributes)
+	ms.writeEventToCache(event)
```
Comment (yycptt) [inline service/history/workflow/mutable_state_impl.go:777]: For later: I think we should allow plugins/components customizing this.

File: service/history/historybuilder/event_store.go [hunk 4]
```diff
@@ -305,7 +305,9 @@ func (b *EventStore) bufferEvent(
 		enumspb.EVENT_TYPE_START_CHILD_WORKFLOW_EXECUTION_INITIATED,
 		enumspb.EVENT_TYPE_SIGNAL_EXTERNAL_WORKFLOW_EXECUTION_INITIATED,
 		enumspb.EVENT_TYPE_UPSERT_WORKFLOW_SEARCH_ATTRIBUTES,
-		enumspb.EVENT_TYPE_WORKFLOW_PROPERTIES_MODIFIED:
+		enumspb.EVENT_TYPE_WORKFLOW_PROPERTIES_MODIFIED,
+		enumspb.EVENT_TYPE_NEXUS_OPERATION_SCHEDULED,
+		enumspb.EVENT_TYPE_NEXUS_OPERATION_CANCEL_REQUESTED:
```
Comment (yycptt) [inline service/history/historybuilder/event_store.go:310]: Replication logic for those events will be in a separate PR?

**Review by yycptt — APPROVED** [review]
The new command handler LGTM! Only skimmed through the actual nexus command processing logic.

File: service/history/historybuilder/event_store.go (hunk 4 above)
Comment (bergundy) [reply service/history/historybuilder/event_store.go:310]: Yeah, all of replication will be done in a separate stage. Both for callbacks and operations.

File: service/history/workflow/mutable_state_impl.go (hunk 2 above)
Comment (bergundy) [reply service/history/workflow/mutable_state_impl.go:431]: Oh thanks, I didn't know we had that.

File: service/history/workflow/mutable_state_impl.go (hunk 3 above)
Comment (bergundy) [reply service/history/workflow/mutable_state_impl.go:777]: I'd like to know why that's important, I don't have the context. But we can discuss when there's a use case.

File: service/history/command_checker.go (hunk 1 above)
Comment (bergundy) [reply service/history/command_checker.go:956]: Seems highly unlikely that this will be overlooked and that we'll add more close commands but I updated the comment with this explanation.

File: plugins/nexusoperations/workflow/commands_test.go [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+		cancelHandler:   cancelHandler,
+	}
+}
+
+func TestHandleScheduleCommand(t *testing.T) {
+	t.Run("feature disabled", func(t *testing.T) {
+		tcx := newTestContext(t, &nexusoperations.Config{
+			Enabled: dynamicconfig.GetBoolPropertyFnFilteredByNamespace(false),
+		})
+		err := tcx.scheduleHandler(tcx.ms, commandValidator{1}, 1, &commandpb.Command{})
```
Comment (MichaelSnowden) [inline plugins/nexusoperations/workflow/commands_test.go:111]: nit: I would make this `commandValidator{maxPayloadSize: 1}` to make it a bit more readable here and elsewhere.

**Review by MichaelSnowden — APPROVED** [review]

**Review by bergundy — COMMENTED** [review]
Thanks for the reviews.
